Hybrid Air Conditioner

Myoldmopar opened the pull request:
@spencerdutton I made this branch from the current state of the prior Hybrid branch but with all the commits squashed down.  This will remove a bunch of history but it will make it load on Github again so that we can review it.  Could you please verify this looks like your current state?

If you had additional changes local on your machine, here is what I would advise:
- Copy the files you have modified locally to a temp folder
- Do a Git Pull to pull the latest code including this new branch
- Check out this new branch locally
- Paste in your changed files
- Commit and push back up, and work from this new branch from now on

Pull request overview
---------------------
@spencerdutton If this works out, can you update this paragraph with a description of the model again since the original PR won't load for me?

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Code style (parentheses padding, variable names)
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check -- verify this
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add rules to spreadsheet
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Documentation changes in place
 - [ ] Changed docs build successfully
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

Myoldmopar committed:
Initial addition of squashed commits

Myoldmopar commented:
@spencerdutton it looks like the primary reason the code is failing to build is because of the change to SetupOutputVariable's interface in #6157.  It should be a pretty straightforward set of changes and then the code can build again.  Let me know if you have trouble with that, or with this new branch/PR.  Once I am sure you are good to go on this new branch, we can begin the review process.

spencerdutton commented:
Ok great I wasn’t aware it wasn’t building. Thanks 

Sent from my iPhone

> On Jan 12, 2018, at 7:39 AM, Edwin Lee <notifications@github.com> wrote:
> 
> @spencerdutton it looks like the primary reason the code is failing to build is because of the change to SetupOutputVariable's interface in #6157. It should be a pretty straightforward set of changes and then the code can build again. Let me know if you have trouble with that, or with this new branch/PR. Once I am sure you are good to go on this new branch, we can begin the review process.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

spencerdutton committed:
Fixed several issues related to scaling and resetting variables.

spencerdutton committed:
Aligned example file with new 6D lookup table format

spencerdutton committed:
Fixed the 6D lookup table issue.

spencerdutton committed:
bug fix

spencerdutton committed:
code tidy

spencerdutton committed:
code tidy of commented parts

spencerdutton committed:
Fixed outputs that were getting zeroed before being reported.

Myoldmopar commented on datasets/InputData_Mixed.csv:
```diff
@@ -0,0 +1,7 @@
+timestamp,Outside Air Temperature (C),Outside Air Relative Humidity (1-100%),Outside Air Humidity Ratio (-),Zone Thermostat Requested Sensible Heat Transfer Rate (kW),Supply Air Temperature (C),Supply Air Relative Humidity (0-100%),Supply Air Humidity Ratio,Water Use (Liters/min),System Electric Power (kW),Operating Mode,Supply Air Mass Flow Rate (kg/s),Supply Air Volume Flow Rate (m3/s),Sensible ...
```
I'm a little confused what this file is used for.  It just has data for a few random August timestamps?  Doesn't seem like an actual "dataset", so if it needs to tag along with the idf example files, it should also go along into testfiles/ and with a more descriptive, specific filename.

spencerdutton replied:
Sorry that was no longer supposed to be in the branch, I removed it.

Myoldmopar commented on src/EnergyPlus/DataZoneEquipment.cc:
```diff
@@ -702,7 +703,9 @@ namespace DataZoneEquipment {
-					} else {
+					} else if (SELECT_CASE_var == "ZONEHVAC:HYBRIDUNITARYHVAC") {
+						ZoneEquipList(ControlledZoneNum).EquipType_Num(ZoneEquipTypeNum) = ZoneHybridEvaporativeCooler_Num;
+					}else {
```
Funny spacing here.

spencerdutton replied:
fixed

Myoldmopar commented on src/EnergyPlus/DataZoneEquipment.hh:
```diff
@@ -94,6 +94,7 @@ namespace DataZoneEquipment {
 	extern int const VRFTerminalUnit_Num;
 	extern int const PurchasedAir_Num;
 	extern int const ZoneEvaporativeCoolerUnit_Num; // #13, last zone equipment type to use zone availability manager. The above list must not change or NumValidSysAvailZoneComponents(13) must also change.
+	extern int const ZoneHybridEvaporativeCooler_Num;// #14, last zone equipment type to use zone availability manager. The above list must not change or NumValidSysAvailZoneComponents(14) must also change.
```
You copy/pasted the line above, but the comment needs to be removed from the line above or this doesn't make sense.

spencerdutton replied:
fixed

Myoldmopar commented on src/EnergyPlus/HybridEvapCoolingModel.cc:
```diff
@@ -0,0 +1,2389 @@
+			Max_Msa = maxMsa;
+			return true;
+		}
+		bool CMode::ValidPointer(int curve_pointer) {
```
What's the point of this function?  Zero is already falsy in C++.

spencerdutton replied:
Well this related to the question later on about what to do if we don't have a valid curve pointer. There are cases where the curve might not be set or required. Like in standby mode for example. So in that case the curve pointerID will be -1 at that point, so we have a check. I'm sure there is a more elegant solution to this, but i'm not loosing sleep over it (except I am ;-).

Myoldmopar commented on src/EnergyPlus/HybridEvapCoolingModel.cc:
```diff
@@ -0,0 +1,2389 @@
+		}
+		Real64 CMode::CalculateCurveVal( Real64 X_1, Real64 X_2, Real64 X_3, Real64 X_4, Real64 X_5, Real64 X_6, int curveType)
+		{
+		// SUBROUTINE INFORMATION:
```
These comments need to be pushed out one more indentation.

spencerdutton replied:
fixed

Myoldmopar commented on src/EnergyPlus/HybridEvapCoolingModel.cc:
```diff
@@ -0,0 +1,2389 @@
+
+				if (ValidPointer(HRsa_curve_pointer))
+				{
+					Y_val = NormalizationReference*CurveValue(HRsa_curve_pointer, X_1, X_2, X_3, X_4, X_5, X_6);//0.01;//
```
Stray comment.

spencerdutton replied:
Fixed

Myoldmopar commented on src/EnergyPlus/HybridEvapCoolingModel.cc:
```diff
@@ -0,0 +1,2389 @@
+					Y_val = Correction*CurveValue(Psa_curve_pointer, X_1, X_2, X_3, X_4, X_5, X_6);
+				}
+				else {
+					Y_val = 0;//or set a more reasonable default
```
This and the below items say "or set a more reasonable default".  So...should you be setting a more reasonable default?  If not, shouldn't the curve just be required?

spencerdutton replied:
Yes thats a good point, I will chat with Jonathan about what we really want to happen if a user doesn't give us a valid curve for power use. Should we just give a warning and return 0 for power in that mode?  Maybe.

Myoldmopar commented on src/EnergyPlus/HybridEvapCoolingModel.cc:
```diff
@@ -0,0 +1,2389 @@
+			RawHR(0)
+		{
+			WarnOnceFlag = false;
+			cp = 1; //kJ/degreesC.kg default value, reset during calculation
```
Specific heat in E+ is commonly held as J/kgK, not kJ.  Unless there is a valid reason to keep it as kJ, it needs to change to J.

spencerdutton replied:
Fixed, all Cp's now in J/kgK

Myoldmopar commented on src/EnergyPlus/HybridEvapCoolingModel.cc:
```diff
@@ -0,0 +1,2389 @@
+		{
+			WarnOnceFlag = false;
+			cp = 1; //kJ/degreesC.kg default value, reset during calculation
+			Lambna = 2260; //(kJ/kg) latent heat of vaporization ePlus should carry this an available variable
```
1. Spelling.  Is this supposed to be lamdba?
2. Yes, E+ has it.  Psychrometrics::PsyHfgAirFnWTdb().  Note that the function accepts two arguments, but the first is unused and will be removed at some point, so just pass temperature as the second argument.

spencerdutton replied:
Great, thanks thats addressed now in the code I use PsyHfgAirFnWTdb whenever I need a lambda,

Myoldmopar commented on src/EnergyPlus/HybridEvapCoolingModel.cc:
```diff
@@ -0,0 +1,2389 @@
+			int MAXIMUM_OPERATIONAL_SETTINGS = 5;
+			ModeCounter = 0;
+			//!!!!!!!!!!!!!!!!!!!!!!!!! this debug code will be removed nearer the code freeze, please don't comment on it, it will be gone.
+			DebugBreak = 12; // remove
```
OK.  I'll trust that it will be gone then...

Myoldmopar commented on src/EnergyPlus/HybridEvapCoolingModel.cc:
```diff
@@ -0,0 +1,2389 @@
+		{
+			for (auto iter = SolutionSpaces.begin(); iter != SolutionSpaces.end(); ++iter) {
+				CModeSolutionSpace* p = (*iter);
+				delete p;
```
I don't like what I'm seeing here.  Calling `delete` is a big red flag here in C++.  My suggestion is to quickly figure out an alternative solution to manual memory management before any additional review.

spencerdutton replied:
This was old obsolete code left over by mistake, i'm sorry. I have pushed up code where that obsolete code is removed. 

In the remainder of the code here is now only the occasional use of "safe" pointers 

std::shared_ptr< CSetting > pCandidateSetting = std::shared_ptr< CSetting >(new CSetting());

jasondegraw replied:
Sorry, but I don't think the alternative code is much better. Why is this needed? There still appears to be manual memory management in various forms throughout this code.

spencerdutton replied:
Yes I had messed up the push up of the code (thanks smartgit) there should be no deletes left if the  code only occasional use of safe pointers that are I understand  also used in the rest of the code base. I found it necessary to use safe pointers because the number of settings the model looks at changes radically depending on how you specify the model, it needs the flexibility afforded by the safe pointers. 

Can you "point" (ha) me to an example of manual management that is not acceptable?

jasondegraw replied:
GitHub is already starting to drag, so I think we're already on our way to another angry unicorn. The use of smart pointers without more justification (why shared instead of unique, for example) is a reddish flag for me, and then were non-trivial destructors that I can no longer find. Maybe those were also obsolete, too? In that case, the branch isn't really ready for review.

spencerdutton replied:
@jasondegraw. The number of potential settings a system can be in can vary dramatically depending on the type of system being modeled. A setting is a unique combination of, the mode the unit is in, the outside air fraction, and supply air mass flow rate. The way this is dealt with in the code is we keep a very large list of all the possible setting this model could be in, that doesn't change once its setup.  Then, when the system is operating one or more of these setting are chosen to be kept in the OperatingSettings list (the setting the unit is actually using this time step). This means we have a single resource that is being referred to in more than one location. So, I feel like the use of a shared smart pointers here is justified, or at least I can not think of a more elegant solution.

Myoldmopar commented on src/EnergyPlus/HybridEvapCoolingModel.cc:
```diff
@@ -0,0 +1,2389 @@
+	
+		}
+
+		int Model::GetID()
```
Although getters and setters are a good idea in most projects, we don't actually need them in E+.  Just make the ID variable a public member and grab it directly.  If this function is intended to do more work before returning the member, such as looking up some other data, then sure, a mining function would be OK.

spencerdutton replied:
ok removed those.

Myoldmopar reviewed (commented):
@spencerdutton The first thing I notice is that you have your idf example files in the datasets/ folder.  These should be in the testfiles/ folder, and you should include `ADD_SIMULATION_TEST` lines for each new example file you add.  This will ensure our CI systems run the example files while it is still on a branch.  The error files for the example files should be very clean, ideally no warnings or errors, but some can be justified.

(Now I am leaving inline review comments.)

OK, so I will say this is overall looking much better.  However, once I saw manual memory management, that was where I stopped.  I can't see how that is necessary in this work and we won't be letting that aspect in.  Can you explain why you went in that direction on this?  Maybe that would help narrow in on a better solution.  I don't see polymorphism at play here, so I think you can probably just change the pointers to actual instances everywhere and be done with that problem.

Anyway, I also left several other comments.  Please address, and push, as soon as possible, and I can make another pass over it.

(I can't mark this as "request changes" since I'm the one who opened this PR, but it should currently be at that state.)

Myoldmopar commented:
(Oh, and CI is also complaining about the license text.  Verify your text matches the other source files **exactly** and that warning should go away.

spencerdutton committed:
removed old memory management code

spencerdutton committed:
removed example file

spencerdutton committed:
removed example file

spencerdutton committed:
removed old input file

spencerdutton committed:
renamed confusingly named variable

jasondegraw commented:
@Myoldmopar I fear the modifications made here are going to be too extensive (again) for GitHub-based review to work. @lgentile has previously suggested that for these situations we should schedule a code review meeting for all the interested parties and just hash it out until we get to a resolution: in or not in. I don't see how this goes in for the release unless we do something other than a GitHub review. I feel like it could angry unicorn on me at any time.

spencerdutton commented:
@jasondegraw.  Can someone please explain what’s the issue here. There are no more unsafe pointers, all the safe pointers are I believe handled appropriately.  I would love to have a code review meeting, I’m available, literally anytime. 

Sent from my iPhone



Sent from my iPhone
> On Feb 11, 2018, at 4:48 PM, Jason DeGraw <notifications@github.com> wrote:
> 
> @Myoldmopar I fear the modifications made here are going to be too extensive (again) for GitHub-based review to work. @lgentile has previously suggested that for these situations we should schedule a code review meeting for all the interested parties and just hash it out until we get to a resolution: in or not in. I don't see how this goes in for the release unless we do something other than a GitHub review. I feel like it could angry unicorn on me at any time.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

spencerdutton committed:
Addressed review comments

Addresses Edwin's comments, changes Cp to the more commonly used units in EP, spacing and other issues.

spencerdutton committed:
Address lambda calculation

I used to have a fixed value for lambda, now I calculated it using Psychrometrics::PsyHfgAirFnWTdb(), as Edwin suggested (thanks)

spencerdutton committed:
Removed getters and setters.

removed at Edwin's request.

spencerdutton committed:
Refactoring to tidy referencing objects.

Myoldmopar commented:
Thanks for continuing to make changes @spencerdutton.  I am looking it over again.  Something funny happened with a commit though.  If you look at the files changed tab, you'll see it looks like some files got fully rewritten.  DataZoneEquipment for example.  Did you make a line endings or broad spacing change that caused this?  https://github.com/NREL/EnergyPlus/pull/6432/commits/27cde653d5a4212d8addb1417b1860529425f1b0 is the diff where it was introduced last night.

spencerdutton committed:
Code refactoring to tidy up how some objects are referenced

spencerdutton committed:
Delete HybridEvapCoolingModel.hh

spencerdutton committed:
Delete HybridEvapCoolingModel.cc

spencerdutton commented:
If I did I didn't intend on doing so, I just meant to add the extra spacing
you asked for. Sorry. :-(

On Mon, Feb 12, 2018 at 12:21 PM, Edwin Lee <notifications@github.com>
wrote:

> Thanks for continuing to make changes @spencerdutton
> <https://github.com/spencerdutton>. I am looking it over again. Something
> funny happened with a commit though. If you look at the files changed tab,
> you'll see it looks like some files got fully rewritten. DataZoneEquipment
> for example. Did you make a line endings or broad spacing change that
> caused this? 27cde65
> <https://github.com/NREL/EnergyPlus/commit/27cde653d5a4212d8addb1417b1860529425f1b0>
> is the diff where it was introduced last night.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NREL/EnergyPlus/pull/6432#issuecomment-365049671>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AJw4Td3EA4c5xIYEDfSLcpmsnjCwBEV7ks5tUJ04gaJpZM4RZ4nx>
> .
>



-- 
Dr Spencer Dutton
Building 90 R 3058
Indoor Environment
Commercial Ventilation and Indoor Air Quality Group
Lawrence Berkeley National Laboratory
1 Cyclotron Rd, Berkeley, California, USA
Tel:  510 4867179
Email: smdutton@lbl.gov

spencerdutton commented:
Also I just checked in some changes a few seconds ago

On Mon, Feb 12, 2018 at 12:25 PM, Spencer Dutton <smdutton@lbl.gov> wrote:

> If I did I didn't intend on doing so, I just meant to add the extra
> spacing you asked for. Sorry. :-(
>
> On Mon, Feb 12, 2018 at 12:21 PM, Edwin Lee <notifications@github.com>
> wrote:
>
>> Thanks for continuing to make changes @spencerdutton
>> <https://github.com/spencerdutton>. I am looking it over again.
>> Something funny happened with a commit though. If you look at the files
>> changed tab, you'll see it looks like some files got fully rewritten.
>> DataZoneEquipment for example. Did you make a line endings or broad spacing
>> change that caused this? 27cde65
>> <https://github.com/NREL/EnergyPlus/commit/27cde653d5a4212d8addb1417b1860529425f1b0>
>> is the diff where it was introduced last night.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/NREL/EnergyPlus/pull/6432#issuecomment-365049671>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AJw4Td3EA4c5xIYEDfSLcpmsnjCwBEV7ks5tUJ04gaJpZM4RZ4nx>
>> .
>>
>
>
>
> --
> Dr Spencer Dutton
> Building 90 R 3058
> Indoor Environment
> Commercial Ventilation and Indoor Air Quality Group
> Lawrence Berkeley National Laboratory
> 1 Cyclotron Rd, Berkeley, California, USA
> Tel:  510 4867179 <(510)%20486-7179>
> Email: smdutton@lbl.gov
>



-- 
Dr Spencer Dutton
Building 90 R 3058
Indoor Environment
Commercial Ventilation and Indoor Air Quality Group
Lawrence Berkeley National Laboratory
1 Cyclotron Rd, Berkeley, California, USA
Tel:  510 4867179
Email: smdutton@lbl.gov

Myoldmopar commented on src/EnergyPlus/HybridUnitaryAirConditioners.cc:
```diff
@@ -0,0 +1,728 @@
+
+				for (int i = 0; i< size; i++)
+				{
+					CSetting *po = &ZoneHybridUnitaryAirConditioner(UnitLoop).CurrentOperatingSettings[i];			
```
This isn't necessary, right?  If you want to create a reference, keep it simple:

```
auto & thisSetting = ZoneHybridUnitaryAirConditioner(UnitLoop).CurrentOperatingSettings[i];
```

And in the string variable names, just use plain C++ concatenation instead of creating a string stream object.

That will make this section a lot cleaner.

spencerdutton replied:
ok done

Myoldmopar commented on src/EnergyPlus/HybridUnitaryAirConditioners.cc:
```diff
@@ -0,0 +1,728 @@
+			}
+		}
+		// this is used excusively for unit testing to get access to the ZoneHybridUnitaryAirConditioner
+		Model* HandleToHybridUnitaryAirConditioner(int UnitNum)
```
This function is just hanging in the global state.  I don't see what you gain by having this pointer function.  Why not just get the `ZoneHybridUnitaryAirConditioner(UnitNum)` variable directly in your unit test??  I think this needs to go.

spencerdutton replied:
ok done

Myoldmopar commented on src/EnergyPlus/HybridUnitaryAirConditioners.cc:
```diff
@@ -0,0 +1,728 @@
+		//_______________________________________________________________________________________________________________________
+		//***************
+		void
+			SimZoneHybridUnitaryAirConditioners(
```
Still some formatting/indentation issues to address before we're all done, but this is looking much better overall.

spencerdutton replied:
ok done

Myoldmopar commented on src/EnergyPlus/HybridEvapCoolingModel.hh:
```diff
@@ -0,0 +1,414 @@
+
+	namespace HybridEvapCoolingModel {
+
+		enum class SYSTEMOUTPUTS { VENTILATION_AIR_V, SUPPLY_MASS_FLOW, SYSTEM_FUEL_USE, SUPPLY_AIR_TEMP, MIXED_AIR_TEMP, SUPPLY_AIR_HR, MIXED_AIR_HR, OSUPPLY_FAN_POWER, OSECOND_FUEL_USE, OTHIRD_FUEL_USE, OWATER_USE, OEXTERNAL_STATIC_PRESSURE};
```
enum classes are always appreciated.

spencerdutton replied:
Is that a positive affirmation? ;-)

jasondegraw replied:
My personal preference would be to camel case (`SystemOutputs`, `Ventilation_Air_V`, etc.), but not a big deal.

Myoldmopar commented on src/EnergyPlus/HybridEvapCoolingModel.hh:
```diff
@@ -0,0 +1,414 @@
+
+			// add other fuels, or change name to be fuel
+			CMode oMode;
+			CSetting& operator=(CSetting other)
```
What's the gain over the default copy operator here?  What makes a custom one necessary?

spencerdutton replied:
it used to have something in it that required it, now it doesn't so I removed it. Thanks

Myoldmopar commented on src/EnergyPlus/HybridEvapCoolingModel.cc:
```diff
@@ -0,0 +1,2383 @@
+			int MAXIMUM_OPERATIONAL_SETTINGS = 5;
+			ModeCounter = 0;
+			//!!!!!!!!!!!!!!!!!!!!!!!!! this debug code will be removed nearer the code freeze, please don't comment on it, it will be gone.
+			DebugBreak = 12; // remove
```
Reminder here to remove this later.

Myoldmopar commented on src/EnergyPlus/HybridEvapCoolingModel.cc:
```diff
@@ -0,0 +1,2383 @@
+		CMode::~CMode()  
+		{
+	
+		}
```
Why have your own explicitly written but empty destructors?

spencerdutton replied:
they used to contain things. removed

Myoldmopar commented on src/EnergyPlus/HybridEvapCoolingModel.cc:
```diff
@@ -0,0 +1,2383 @@
+			return (P * W / (0.62198 + W));
+		}
+		
+		Real64 Model::Sat_press(Real64 Tdb)
```
We should already have this in Psychrometrics.hh, we won't want to have this repeated here.

spencerdutton replied:
done

jasondegraw replied:
The commented out functions here should be deleted.

Myoldmopar commented on src/EnergyPlus/HybridEvapCoolingModel.cc:
```diff
@@ -0,0 +1,2383 @@
+			Real64 Pws = Sat_press(Tdb);
+			Real64 Hum_rat = 0.62198 * RH * Pws / (P - RH * Pws);   // Equation 22, 24, p6.8
+			return Hum_rat;
+		}
```
Also should be available in Psychrometrics.hh.

spencerdutton replied:
done

Myoldmopar commented on src/EnergyPlus/HybridEvapCoolingModel.cc:
```diff
@@ -0,0 +1,2383 @@
+			{
+				if (OutletW < Wsa)
+				{
+					ShowWarningError("Humidity ratio exceeded realistic range, check performance curve");
```
Is this warning going to be thrown many times?  If so, it will make the error file unreadable, so you may need to trap for multiple calls to it.

spencerdutton replied:
fixed

Myoldmopar commented on src/EnergyPlus/HybridEvapCoolingModel.cc:
```diff
@@ -0,0 +1,2383 @@
+		{
+			Real64 OutletRHtest = PsyRhFnTdbWPb(T, W, P); //could also use outlet pressure instead of fixed
+			Real64 OutletW = PsyWFnTdbRhPb(T, OutletRHtest, P); 
+			if (OutletRHtest == 1)
```
What is the == 1 implying here?  Is this a boolean?  Or a real?  Or is it actually a meaningful integer where 1 should be some parameter and not a magic number?

spencerdutton replied:
it was to check to see if it had topped out, I refactored this.

Myoldmopar commented on src/EnergyPlus/HybridEvapCoolingModel.cc:
```diff
@@ -0,0 +1,2383 @@
+			if (StepIns.RHosa > 1) 
+			{ 
+				ShowSevereError("Unitary hybrid system error, required relative humidity value 0-1, called in object" + ObjectID + ".Check inputs" ); 
+				return -1;
```
I would've sworn earlier I saw you were using try/catch, so I'd expect you to throw a meaningful severe error here, then raise an exception so that the caller can throw the final fatal.  Either that or just throw the fatal error here.

spencerdutton replied:
I think I addressed this, take a look?

Myoldmopar commented on src/EnergyPlus/HybridEvapCoolingModel.cc:
```diff
@@ -0,0 +1,2383 @@
+							//Return Air Relative Humidity(0 - 100 % ) //Return Air Humidity Ratio(g / g)
+							if (MeetsSupplyAirRHOC(Wsa))
+							{
+				
```
We could eliminate all of the `(*iteratorSetting)->` lines below if this was changed to a range-based for loop.

```
for (auto & thisSetting: Settings) {
  thisSetting.Mixed_Air_Temperature = Tma;
  //...etc.
}
```

I think that would be a simple change to make that would make this code much more readable than constantly dereferencing the pointer.

spencerdutton replied:
awesome thanks, done

Myoldmopar reviewed (commented):
@spencerdutton OK, another round of comments.  I really think that this is starting to make good progress.  Please address these, and figure out what went wrong with the two files that show full diffs here on the Files Changed tab.

Once you address these comments, I'll pull and build and start testing it out.  I'll also work on getting the docs pulled in.

Myoldmopar commented on src/EnergyPlus/HybridEvapCoolingModel.cc:
```diff
@@ -0,0 +1,2383 @@
+							//Return Air Relative Humidity(0 - 100 % ) //Return Air Humidity Ratio(g / g)
+							if (MeetsSupplyAirRHOC(Wsa))
+							{
+				
```
@spencerdutton Why does this have to be a shared pointer?  What are you sharing across the multiple references to the same memory?  Unless you have a solid reason for doing it, please make the following changes to avoid this being a red flag:

- Change the definition to simply a list of CSettings instead of a list of pointers
- In this section, just instantiate a new CSettings object and assign the values
- Change any code that attempts to dereference the Settings variable to use standard object dot notation

Then we can move fully away from any pointer concerns with the team.

spencerdutton replied:
done

spencerdutton committed:
address Edwin's comments

spencerdutton committed:
Updated unit test

spencerdutton committed:
Update unit test

Myoldmopar committed:
Changed line endings in DataZoneEquipment files; should eliminate stray diffs

Myoldmopar commented:
@spencerdutton I don't know what editor you are using for your C++ changes, but something got botched with the line endings on DataZoneEquipment files.  I corrected it and pushed back up.  Please make a pull now so that you don't end up having merge issues later.

Myoldmopar committed:
Cleanup some compiler warnings

Myoldmopar commented:
@spencerdutton I'm a little confused about the IDF file you added to the datasets/ folder.  The name of the file says UnitTest.  But it looks like an entire IDF.  And I don't see a new example file added to the testfiles/ folder.

You do have unit tests written, covering multiple operating modes, and that is good.  But that is totally separate from that IDF, right?  That IDF should really just be put into testfiles/ and included in the release as a user example file.

I'm moving forward with that assumption.  

I've cleaned up some compiler warnings, but there are still likely to be some.  The declarations of class member variables isn't fully aligned with the order in the initializer list in the constructor, so the compiler warns about it.  I've cleaned up some but it's only Clang that is warning about it, and I'm on Ubuntu, so I can't see them locally.  If it turns out to be a game of whack-a-mole I can start a build on another machine.

I'm going to move that IDF into testfiles/ and add the CMake declaration for it.  Please pull before you make any more changes.

Myoldmopar committed:
Move example file, clean up some compiler warnings

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into Hybrid_New_Branch

Myoldmopar committed:
Cleanup more compiler issues

Myoldmopar committed:
Fix initialization bug

Myoldmopar commented:
Alright I just pushed up a collection of fixes.  This *should* clean up the compiler warnings, of which there were several.  I also moved the IDF to the testfiles/ folder.  I also cleaned up an initialization bug (please see dacaf7a and verify).  At this point, the IDF is at least starting to run, but very slow.  The slow model is amplified by my debug version of the build.  I'm building a release optimized version so that maybe it will get through it quicker.

Make sure you pull these changes down if you make any more changes.  I'll keep testing now that things are cleaned up further.

Myoldmopar committed:
Cleanup IDF errors/warnings, rename hybrid error code output variable

Myoldmopar committed:
Continued cleaning up input file

Myoldmopar commented on src/EnergyPlus/HybridEvapCoolingModel.cc:
```diff
@@ -0,0 +1,2368 @@
+			//assess the sixe of the solution space to make sure it holds at least 1 combination
+			int solution_map_sizeX = solutionspace.PointX.size();
+			//Check the normalization reference is set if not set a default
+			Real64 NormalizationReference = GetNormalPoint(1);// assumes the model has at least 1 curve and that all are set the same, this is explained in the IO reference
```
I need some clarification on this.  What do you mean it assumes the model has at least one curve?  The IDF could be littered with curves!  And you are calling it to get a normalization reference for the first one??  I don't understand this.

spencerdutton commented:
No the unit test idf is part the unit test, the hybrid model is so large with the 6d lookup tables it made sense to have it as an external file (I believe we talked about this at the time). We have a couple of examples that could be the example file, of which the unit test example is one, but we would change the name obviously. 

Sent from my iPhone

> On Feb 14, 2018, at 3:31 AM, Edwin Lee <notifications@github.com> wrote:
> 
> @spencerdutton I'm a little confused about the IDF file you added to the datasets/ folder. The name of the file says UnitTest. But it looks like an entire IDF. And I don't see a new example file added to the testfiles/ folder.
> 
> You do have unit tests written, covering multiple operating modes, and that is good. But that is totally separate from that IDF, right? That IDF should really just be put into testfiles/ and included in the release as a user example file.
> 
> I'm moving forward with that assumption.
> 
> I've cleaned up some compiler warnings, but there are still likely to be some. The declarations of class member variables isn't fully aligned with the order in the initializer list in the constructor, so the compiler warns about it. I've cleaned up some but it's only Clang that is warning about it, and I'm on Ubuntu, so I can't see them locally. If it turns out to be a game of whack-a-mole I can start a build on another machine.
> 
> I'm going to move that IDF into testfiles/ and add the CMake declaration for it. Please pull before you make any more changes.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

spencerdutton commented:
Yes that’s not clear, I mean it assumes that specified in the unitary hybrid model is at least 1 performance table.. which is I guess a reasonable assumption, because without the tables you have no model. 

Sent from my iPhone

> On Feb 14, 2018, at 4:40 AM, Edwin Lee <notifications@github.com> wrote:
> 
> @Myoldmopar commented on this pull request.
> 
> In src/EnergyPlus/HybridEvapCoolingModel.cc:
> 
> > +			// METHODOLOGY EMPLOYED:
> +			// Uses the relevant lookup take to specify the parameters, or uses default conditions.
> +			// In setting the supply air temperature for now just use return air future improvement will use look up table
> +
> +
> +			// REFERENCES:
> +			// na
> +
> +			// Using/Aliasing
> +			// using the pointer to the first of the operating modes (which is always the standby mode by definition.
> +			//Get a handel to the map of possible solutions (valid combinations of OSA fraction and Msa)
> +			CModeSolutionSpace solutionspace = Mode0.sol;
> +			//assess the sixe of the solution space to make sure it holds at least 1 combination
> +			int solution_map_sizeX = solutionspace.PointX.size();
> +			//Check the normalization reference is set if not set a default
> +			Real64 NormalizationReference = GetNormalPoint(1);// assumes the model has at least 1 curve and that all are set the same, this is explained in the IO reference
> I need some clarification on this. What do you mean it assumes the model has at least one curve? The IDF could be littered with curves! And you are calling it to get a normalization reference for the first one?? I don't understand this.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

jasondegraw commented on src/EnergyPlus/HybridEvapCoolingModel.cc:
```diff
@@ -0,0 +1,2368 @@
+			Real64 value = 0;
+
+			for (auto & thisOperatingSettings : CurrentOperatingSettings) {
+				switch (val)
```
I don't understand why this switch is inside the loop when the switch argument is a function argument.

spencerdutton replied:
The CurrentOperatingSettings is the list of settings the unit is in for that timestep. For longer timesteps the model can be in multiple settings, like a cooling setting for 40% of the time, and off for 60% of the time for example. So firstly we iterate thought these settings, and then depending on the output variable we are trying to calculate, we use the switch to change the averaging calculation method. Average mass flow rate for example is not the same calculation as average supply air temperature. 

The switch changes how we do the averaging of the variables. Extensive variables including: supply air mass flow rate, fuel uses, and water use, are one approach, Intensive variables take a different route. Now as there are only two different types of variables im sure we could have come up with a smarter way to do this, but for now I can't think of one.

Myoldmopar commented:
> Yes that’s not clear, I mean it assumes that specified in the unitary hybrid model is at least 1 performance table.. which is I guess a reasonable assumption, because without the tables you have no model.

But, it's calling the curve normalization function with an index of 1.  Curve index 1 could have nothing do do with the hybrid model.  It could be the boiler performance on a plant object.  Or a pressure curve in a pressure based simulation.  Or something completely different.  Using a 1 there just can't be the right option, right?  You need to pass something meaningful in there.  You should pass in the index of a curve that you are using for the hybrid model.  If I'm missing something about the functionality here, please clarify.  

On a brighter note, this branch is now showing diffs in only two files and they are completely expected, so this code isn't affecting any other part of E+.

spencerdutton commented:
Yes that was dumb I’ll fix that sorry. 

Sent from my iPhone

> On Feb 14, 2018, at 9:48 AM, Edwin Lee <notifications@github.com> wrote:
> 
> Yes that’s not clear, I mean it assumes that specified in the unitary hybrid model is at least 1 performance table.. which is I guess a reasonable assumption, because without the tables you have no model.
> 
> But, it's calling the curve normalization function with an index of 1. Curve index 1 could have nothing do do with the hybrid model.  It could be the boiler performance on a plant object. Or a pressure curve in a pressure based simulation. Or something completely different. Using a 1 there just can't be the right option, right? You need to pass something meaningful in there. You should pass in the index of a curve that you are using for the hybrid model. If I'm missing something about the functionality here, please clarify.
> 
> On a brighter note, this branch is now showing diffs in only two files and they are completely expected, so this code isn't affecting any other part of E+.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

Myoldmopar commented:
OK cool.  And as for the unit test/example file thing, if it is a file needed for unit test only, we should probably put it close to the unit tests themselves.  The `datasets/` folder is for actual equipment specs that can be dropped into other IDFs, not whole IDFs themselves.  I'll move it over there and make sure the unit test is still passing.  If you can add some of your functioning example files (1-2 of them) into `testfiles/`and make sure they get added (alphabetically) to the entries in `testfiles/CMakeLists.txt`, that would be great.

The Windows 64 build seemed to fall on it's face, but the 32 bit build was fine and the other builds seem fine, so I think it was a fluke and I restarted that build.  We'll see.

jasondegraw commented on src/EnergyPlus/HybridEvapCoolingModel.hh:
```diff
@@ -0,0 +1,373 @@
+			CMode();
+			
+			//finish init above
+			~CMode() {};                  // destructor
```
Is this empty destructor really needed?

Myoldmopar replied:
Should not be needed.  Should remove.

spencerdutton replied:
I removed

jasondegraw commented on src/EnergyPlus/HybridEvapCoolingModel.hh:
```diff
@@ -0,0 +1,373 @@
+
+		enum class SYSTEMOUTPUTS { VENTILATION_AIR_V, SUPPLY_MASS_FLOW, SYSTEM_FUEL_USE, SUPPLY_AIR_TEMP, MIXED_AIR_TEMP, SUPPLY_AIR_HR, MIXED_AIR_HR, OSUPPLY_FAN_POWER, OSECOND_FUEL_USE, OTHIRD_FUEL_USE, OWATER_USE, OEXTERNAL_STATIC_PRESSURE};
+
+		class CModeSolutionSpace
```
If everything in these classes is public, it'd be better that these be structs. That way everything is public be default, no labels needed.

spencerdutton replied:
Im not following but if its not a big deal Ill keep it as a enum class

jasondegraw replied:
This comment was not a reference to an enum class, I'm talking about all the classes that have no private members. If everything is public, it's better to use a struct and then the `public:`, etc. labels are no longer needed.

jasondegraw commented on src/EnergyPlus/HybridEvapCoolingModel.hh:
```diff
@@ -0,0 +1,373 @@
+			CSetting oStandBy;
+
+			//std::list <std::shared_ptr< CSetting >> Settings;
+			std::list <CSetting> Settings;
```
I definitely prefer `std::vector` as opposed to `std::list` here unless there is a really compelling reason to go with list.

Myoldmopar replied:
Yeah, why is this a list and not a vector?  @spencerdutton, was there a reason for list?  Because if not, please make it a vector.

spencerdutton replied:
changed it to vector thanks

jasondegraw commented on src/EnergyPlus/HybridUnitaryAirConditioners.cc:
```diff
@@ -0,0 +1,694 @@
+					CheckZoneHybridEvapName(CompNum) = false;
+				}
+			}
+			try
```
@Myoldmopar are we following this pattern anywhere else?

Myoldmopar replied:
We aren't, but I'm not opposed to this being a demonstration, and using it as a talking point when any new feature is being designed.

spencerdutton replied:
ok ill keep it.

jasondegraw commented on tst/EnergyPlus/unit/UnitaryHybridAirConditioner.unit.cc:
```diff
@@ -0,0 +1,388 @@
+		 ExternalStaticPressure = pZoneHybridUnitaryAirConditioner->ExternalStaticPressure;	   //
+		// checks 
+		
+		if ((roundf(returnOSAF * 100) / 100)  != 1) Fail = true;
```
One-line if statements are very easy to mess up, can we add brackets to all of these?

spencerdutton replied:
addressed thanks.

jasondegraw commented on tst/EnergyPlus/unit/UnitaryHybridAirConditioner.unit.cc:
```diff
@@ -0,0 +1,388 @@
+	}
+
+
+	TEST_F(EnergyPlusFixture, Test_UnitaryHybridAirConditioner_Unittest) {
```
This unit test confuses me. There's a lot of code with no asserts or expects.

spencerdutton replied:
There are checks after each call to the doStep method. Those checks will determine if the test fails or not. If the test fails there is one assert at the bottom. I believe I copied this approach from a different unit test. Alternatively there could be an assert after each check or group of checks. For some reason my ASSERT at the end was missing you are right, I added that back.

spencerdutton replied:
hopefully thats up now.

jasondegraw reviewed (changes requested):
There are a lot of formatting issues, which may not matter here pretty soon after the freeze, but more importantly I still have concerns about the use of `std::list` and the overall complexity of what is going on. It doesn't look to me like the use of list is justified. There's some commented code that needs to be removed, I've commented on one.

Myoldmopar commented:
Also, @spencerdutton, could you provide links to the docs for the IO and Enginering Reference?

spencerdutton commented:
Sure hopefully this link works. I assumed you already had seen this.

https://docs.google.com/document/d/1h-MtZO7qrEHFAP8ymUAE15fm_7X-
FRtwiIbHd1qmLi0/edit
https://docs.google.com/document/d/1gOicuX5SDOlbMtXNgISqE9DeKKdb-n_Huvbpsel0GKI/edit

On Wed, Feb 14, 2018 at 11:56 AM, Edwin Lee <notifications@github.com>
wrote:

> Also, @spencerdutton <https://github.com/spencerdutton>, could you
> provide links to the docs for the IO and Enginering Reference?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NREL/EnergyPlus/pull/6432#issuecomment-365726424>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AJw4TQlC10-dvkHEaEbfizcwt--_UNmGks5tUzqGgaJpZM4RZ4nx>
> .
>



-- 
Dr Spencer Dutton
Building 90 R 3058
Indoor Environment
Commercial Ventilation and Indoor Air Quality Group
Lawrence Berkeley National Laboratory
1 Cyclotron Rd, Berkeley, California, USA
Tel:  510 4867179
Email: smdutton@lbl.gov

spencerdutton committed:
Update Unit test to address Jason's comments

Myoldmopar committed:
Use C++latest to get fallthrough

Myoldmopar committed:
Comment out fallthroughs; it throws tons of warnings on my machine again, but works

Myoldmopar committed:
Merge remote-tracking branch 'origin/HotFixVSCompilerFlag' into Hybrid_New_Branch

Myoldmopar committed:
Merge branch 'Hybrid_New_Branch' of github.com:NREL/EnergyPlus into Hybrid_New_Branch

Myoldmopar committed:
Add docs for hybrid model, io and eng ref

Myoldmopar commented:
@spencerdutton OK I added the LaTeX documentation, pulling it essentially verbatim from the Google Docs, except where I found a typo.  They build OK locally and the results look good.  With those in place, and the fallthrough issue eliminated from this branch, this should be building well once again.  I'll try to go back through the comments here and list out any remaining items from @jasondegraw or myself that need to be addressed.

Myoldmopar committed:
Fix license year in unit test file

Myoldmopar commented on tst/EnergyPlus/unit/UnitaryHybridAirConditioner.unit.cc:
```diff
@@ -0,0 +1,433 @@
+
+	TEST_F(EnergyPlusFixture, Test_UnitaryHybridAirConditioner_Unittest) {
+		using namespace InputProcessor;
+		std::vector<std::string> snippet = getAllLinesInFile2(configured_source_directory() + "/datasets/UnitaryHybridUnitTest_MuntersEPX5000.idf");
```
This line needs to be changed now, as the test file here is in the unit test directory, not in `datasets/`.

spencerdutton replied:
this is addressed, at least I think it is, i had some trouble pushing it up.

Myoldmopar commented:
I just fixed the license year issue.  CI should be pretty clean with 270e0d2.

@mjwitte can you provide final comments on the IDD changes?  

@spencerdutton there are a few comments from @jasondegraw that are still not addressed.  They aren't major changes, so please get those addressed and push them up.  

On a positive note, the example file that is currently up there is covering over 90% of the lines of code in your two new cpp files and the header file, which is awesome, considering that a vast majority of the remaining uncovered lines are erroneous code paths that are sometimes pretty difficult to hit with example files.  The unit test coverage results are funny, and I think it's because the unit test is pointing at the wrong input file.  I'm going to change that quickly so that hopefully by morning, CI is **very** happy here.

spencerdutton commented:
OK thank you I will address those this evening. The Hybrid_new branch built
on my machine ! thank you. :-)

Spencer


On Thu, Feb 15, 2018 at 5:56 PM, Edwin Lee <notifications@github.com> wrote:

> I just fixed the license year issue. CI should be pretty clean with
> 270e0d2
> <https://github.com/NREL/EnergyPlus/commit/270e0d2d5378e4bfe89aab09bafbf557c9cf66a4>
> .
>
> @mjwitte <https://github.com/mjwitte> can you provide final comments on
> the IDD changes?
>
> @spencerdutton <https://github.com/spencerdutton> there are a few
> comments from @jasondegraw <https://github.com/jasondegraw> that are
> still not addressed. They aren't major changes, so please get those
> addressed and push them up.
>
> On a positive note, the example file that is currently up there is
> covering over 90% of the lines of code in your two new cpp files and the
> header file, which is awesome, considering that a vast majority of the
> remaining uncovered lines are erroneous code paths that are sometimes
> pretty difficult to hit with example files. The unit test coverage results
> are funny, and I think it's because the unit test is pointing at the wrong
> input file. I'm going to change that quickly so that hopefully by morning,
> CI is *very* happy here.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NREL/EnergyPlus/pull/6432#issuecomment-366123300>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AJw4TTbGljXoy2mr1pZil1Pili-bcMbAks5tVOA5gaJpZM4RZ4nx>
> .
>



-- 
Dr Spencer Dutton
Building 90 R 3058
Indoor Environment
Commercial Ventilation and Indoor Air Quality Group
Lawrence Berkeley National Laboratory
1 Cyclotron Rd, Berkeley, California, USA
Tel:  510 4867179
Email: smdutton@lbl.gov

spencerdutton committed:
Address comments

spencerdutton committed:
fixed file location

Myoldmopar committed:
Enable hybrid unit test; clean up configured function build error

Myoldmopar committed:
Merge branch 'Hybrid_New_Branch' of github.com:NREL/EnergyPlus into Hybrid_New_Branch

Myoldmopar committed:
Add preceding slash to filepath in unit test

Myoldmopar committed:
Use CurveID; clean up unit test

Myoldmopar committed:
Fix curve vs table index; Seeley example file now works

Myoldmopar commented:
I am expecting the unit test to pass now.  I am also expecting the build warning for unused parameter to be gone now.  I have tested two example files, and while I haven't rigored over the outputs, they do complete successfully, and the integration test (and now the unit test) covers a substantial about of the new source code.  I have pushed the LaTeX docs and they are building properly.  @spencerdutton has addressed @jasondegraw's serious concerns as far as I can tell. I fixed a significant issue of my own making in the ConfiguredFunctions source code file I added like a long time ago.  I just looked over the IDD and frack there are a few tab characters I'll remove now, but otherwise no red flags.

I intend on merging this if CI is happy.  We will have time to review the guts a bit more over the next couple weeks if something comes up.

Myoldmopar committed:
Remove tabs from idd

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -34955,6 +34955,4645 @@ ZoneHVAC:EvaporativeCoolerUnit,
+    A9, \field Return Air Node Name
+        \type node
+        \note Return air node for the hybrid unit must be a zone exhaust node.
+   A10, \field Outside Air Node Name
```
This should be "Outdoor" for consistency, also in the note below.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -34955,6 +34955,4645 @@ ZoneHVAC:EvaporativeCoolerUnit,
+   A11, \field Supply Air Node Name
+        \type node
+        \note Supply air node for the hybrid unit must be a zone air inlet node.
+   A12, \field Relief Node Name
```
Relief Node vs Return Node was never clarified - lots of discussion in the google doc but no conclusion.  EnergyPlus assumes that ZoneHVAC units will be balanced - one supply node with equal flow out of the zone at an exhaust node. The system model can internally decide how much of that is recirculated and how much is relief.  I thought the discussion was moving toward dropping this node, but it's still here.

spencerdutton replied:
I wasn't sure if the decision was to drop the node. Im fine with doing that if thats what we conclude is the best thing to do, but Im just a little worried about timing. I'll email.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -34955,6 +34955,4645 @@ ZoneHVAC:EvaporativeCoolerUnit,
+   A10, \field Outside Air Node Name
+        \type node
+        \note Outside air node for the hybrid unit must be an outdoor air node.
+   A11, \field Supply Air Node Name
```
\required-field

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -34955,6 +34955,4645 @@ ZoneHVAC:EvaporativeCoolerUnit,
+        \note Relief node for the hybrid unit must be a zone exhaust node, unless flow is being balanced elsewhere.
+    N1, \field System Maximum Supply Air Flow Rate
+        \type real
+        \minimum> 0
```
Any field with a minimum> 0 and no default should be a \required-field, otherwise if it's left blank it will error out.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -34955,6 +34955,4645 @@ ZoneHVAC:EvaporativeCoolerUnit,
+        \units m3/s
+    N2, \field External Static Pressure at System Maximum Supply Air Flow Rate
+        \type real
+        \minimum> 0
```
\required-field

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -34955,6 +34955,4645 @@ ZoneHVAC:EvaporativeCoolerUnit,
+        \object-list MultiVariateTables
+        \note Enter the name of the System Water Use Lookup Table for Mode 1.
+        \note If this field is blank, Mode 1 does not consume water.
+    N8, \field Mode 1 Minimum Outside Air Temperature
```
All of these should be "Outdoor" instead of "Outside".

spencerdutton committed:
Address idd and test script issues.

spencerdutton committed:
Address Cmake mistake

spencerdutton committed:
remove test filter

Myoldmopar commented:
This is back down to 2 actual regressions, with the other 4 coming from the Kiva merge.

Myoldmopar commented:
OK, I feel comfortable with this dropping in now.  The diffs that are related to this branch are meaningful, the other diffs aren't related.  I believe this has pretty well addressed the review comments.  If there is anything lingering, we can file a followup issue to address in the bug fix period.

spencerdutton commented:
Wow that’s really exciting 

Sent from my iPhone

> On Feb 17, 2018, at 7:34 AM, Edwin Lee <notifications@github.com> wrote:
> 
> Merged #6432.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>
